File: web/src/lib/types/disk/disk.ts
import { z } from 'zod';

export const SmartNVMESchema = z.object({
	device: z.string(),
	criticalWarning: z.string(),
	criticalWarningState: z.object({
		availableSpare: z.union([z.number(), z.string()]),
		temperature: z.union([z.number(), z.string()]),
		deviceReliability: z.union([z.number(), z.string()]),
		readOnly: z.union([z.number(), z.string()]),
		volatileMemoryBackup: z.union([z.number(), z.string()])
	}),
	temperature: z.number(),
	availableSpare: z.number(),
	availableSpareThreshold: z.number(),
	percentageUsed: z.number(),
	dataUnitsRead: z.number(),
	dataUnitsWritten: z.number(),
	hostReadCommands: z.number(),
	hostWriteCommands: z.number(),
	controllerBusyTime: z.number(),
	powerCycles: z.number(),
	powerOnHours: z.number(),
	unsafeShutdowns: z.number(),
	mediaErrors: z.number(),
	errorInfoLogEntries: z.number(),
	warningCompositeTempTime: z.number(),
	temperature1TransitionCnt: z.number(),
	temperature2TransitionCnt: z.number(),
	totalTimeForTemperature1: z.number(),
	totalTimeForTemperature2: z.number()
});

export const SmartCtlSchema = z.object({
	json_format_version: z.array(z.number()),
	smartctl: z.object({
		version: z.array(z.number()),
		pre_release: z.boolean(),
		svn_revision: z.string(),
		platform_info: z.string(),
		build_info: z.string().optional(),
		argv: z.array(z.string()),
		drive_database_version: z
			.object({
				string: z.string()
			})
			.optional(),
		exit_status: z.number()
	}),
	local_time: z.object({
		time_t: z.number(),
		asctime: z.string()
	}),
	device: z.object({
		name: z.string(),
		info_name: z.string(),
		type: z.string(),
		protocol: z.string()
	}),
	smart_status: z.object({
		passed: z.boolean()
	}),
	power_on_time: z.object({
		hours: z.number()
	}),
	power_cycle_count: z.number(),
	temperature: z.object({
		current: z.number()
	}),
	ata_smart_attributes: z.object({
		revision: z.number(),
		table: z.array(
			z.object({
				id: z.number(),
				name: z.string(),
				value: z.number(),
				worst: z.number(),
				thresh: z.number(),
				when_failed: z.string(),
				flags: z.object({
					value: z.number(),
					string: z.string(),
					prefailure: z.boolean(),
					updated_online: z.boolean(),
					performance: z.boolean(),
					error_rate: z.boolean(),
					event_count: z.boolean(),
					auto_keep: z.boolean()
				}),
				raw: z.object({
					value: z.number(),
					string: z.string()
				})
			})
		)
	})
});

export const PartitionSchema = z.object({
	name: z.string(),
	usage: z.string(),
	size: z.number()
});

export const DiskInfoSchema = z.array(
	z.object({
		device: z.string(),
		type: z.string(),
		usage: z.string(),
		size: z.number(),
		gpt: z.boolean(),
		model: z.string(),
		serial: z.string(),
		smartData: z.union([SmartNVMESchema, SmartCtlSchema, z.null()]).optional(),
		wearOut: z.union([z.number(), z.string()]).optional(),
		partitions: z.array(PartitionSchema).optional().default([])
	})
);

export type SmartNVME = z.infer<typeof SmartNVMESchema>;
export type DiskInfo = z.infer<typeof DiskInfoSchema>;
export type Partition = z.infer<typeof PartitionSchema>;

export interface Disk {
	Device: string;
	Type: string;
	Usage: string;
	Size: number;
	GPT: boolean;
	Model: string;
	Serial: string;
	'S.M.A.R.T.': string;
	Wearout: string | number | undefined;
	Partitions: Partition[];
}

File: internal/interfaces/services/disk/disk.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package diskServiceInterfaces

type Partition struct {
	UUID  string `json:"uuid"`
	Name  string `json:"name"`
	Usage string `json:"usage"`
	Size  uint64 `json:"size"`
}

type Disk struct {
	UUID       string      `json:"uuid"`
	Device     string      `json:"device"`
	Type       string      `json:"type"`
	Usage      string      `json:"usage"`
	Size       uint64      `json:"size"`
	Model      string      `json:"model"`
	Serial     string      `json:"serial"`
	GPT        bool        `json:"gpt"`
	SmartData  any         `json:"smartData"`
	WearOut    string      `json:"wearOut"`
	Partitions []Partition `json:"partitions"`
}

type NvmeCriticalWarningState struct {
	AvailableSpare       int `json:"availableSpare"`
	Temperature          int `json:"temperature"`
	DeviceReliability    int `json:"deviceReliability"`
	ReadOnly             int `json:"readOnly"`
	VolatileMemoryBackup int `json:"volatileMemoryBackup"`
}

type SMARTNvme struct {
	Device                    string                   `json:"device"`
	CriticalWarning           string                   `json:"criticalWarning"`
	CriticalWarningState      NvmeCriticalWarningState `json:"criticalWarningState"`
	Temperature               int                      `json:"temperature"`
	AvailableSpare            int                      `json:"availableSpare"`
	AvailableSpareThreshold   int                      `json:"availableSpareThreshold"`
	PercentageUsed            int                      `json:"percentageUsed"`
	DataUnitsRead             int                      `json:"dataUnitsRead"`
	DataUnitsWritten          int                      `json:"dataUnitsWritten"`
	HostReadCommands          int                      `json:"hostReadCommands"`
	HostWriteCommands         int                      `json:"hostWriteCommands"`
	ControllerBusyTime        int                      `json:"controllerBusyTime"`
	PowerCycles               int                      `json:"powerCycles"`
	PowerOnHours              int                      `json:"powerOnHours"`
	UnsafeShutdowns           int                      `json:"unsafeShutdowns"`
	MediaErrors               int                      `json:"mediaErrors"`
	ErrorInfoLogEntries       int                      `json:"errorInfoLogEntries"`
	WarningCompositeTempTime  int                      `json:"warningCompositeTempTime"`
	ErrorCompositeTempTime    int                      `json:"errorCompositeTempTime"`
	Temperature1TransitionCnt int                      `json:"temperature1TransitionCnt"`
	Temperature2TransitionCnt int                      `json:"temperature2TransitionCnt"`
	TotalTimeForTemperature1  int                      `json:"totalTimeForTemperature1"`
	TotalTimeForTemperature2  int                      `json:"totalTimeForTemperature2"`
}

type SmartData struct {
	JSONFormatVersion []int        `json:"json_format_version"`
	Smartctl          SmartctlInfo `json:"smartctl"`
	LocalTime         LocalTime    `json:"local_time"`
	Device            DeviceInfo   `json:"device"`
	SmartStatus       SmartStatus  `json:"smart_status"`
	PowerOnTime       PowerOnTime  `json:"power_on_time"`
	PowerCycleCount   int          `json:"power_cycle_count"`
	Temperature       Temperature  `json:"temperature"`

	ATASmartAttributes  *ATASmartAttributes  `json:"ata_smart_attributes,omitempty"`
	SCSISmartAttributes *SCSISmartAttributes `json:"scsi_smart_attributes,omitempty"`
}

type SmartctlInfo struct {
	Version              []int             `json:"version"`
	PreRelease           bool              `json:"pre_release"`
	SVNRevision          string            `json:"svn_revision"`
	PlatformInfo         string            `json:"platform_info"`
	BuildInfo            string            `json:"build_info"`
	Argv                 []string          `json:"argv"`
	DriveDatabaseVersion DriveDatabaseInfo `json:"drive_database_version"`
	ExitStatus           int               `json:"exit_status"`
}

type DriveDatabaseInfo struct {
	String string `json:"string"`
}

type LocalTime struct {
	TimeT   int    `json:"time_t"`
	Asctime string `json:"asctime"`
}

type DeviceInfo struct {
	Name     string `json:"name"`
	InfoName string `json:"info_name"`
	Type     string `json:"type"`
	Protocol string `json:"protocol"`
}

type SmartStatus struct {
	Passed bool `json:"passed"`
}

type PowerOnTime struct {
	Hours int `json:"hours"`
}

type Temperature struct {
	Current int `json:"current"`
}

type ATASmartAttributes struct {
	Revision int                 `json:"revision"`
	Table    []ATASmartAttribute `json:"table"`
}

type ATASmartAttribute struct {
	ID         int            `json:"id"`
	Name       string         `json:"name"`
	Value      int            `json:"value"`
	Worst      int            `json:"worst"`
	Thresh     int            `json:"thresh"`
	WhenFailed string         `json:"when_failed"`
	Flags      AttributeFlags `json:"flags"`
	Raw        RawData        `json:"raw"`
}

type AttributeFlags struct {
	Value         int    `json:"value"`
	String        string `json:"string"`
	Prefailure    bool   `json:"prefailure"`
	UpdatedOnline bool   `json:"updated_online"`
	Performance   bool   `json:"performance"`
	ErrorRate     bool   `json:"error_rate"`
	EventCount    bool   `json:"event_count"`
	AutoKeep      bool   `json:"auto_keep"`
}

type RawData struct {
	Value  int64  `json:"value"`
	String string `json:"string"`
}

type SCSISmartAttributes struct {
	Temperature int `json:"scsi_temperature,omitempty"`
}

type DiskServiceInterface interface {
	GetDiskDevices() ([]Disk, error)
	GetSmartData(disk DiskInfo) (any, error)
	GetWearOut(disk any) (float64, error)
	GetDiskSize(device string) (uint64, error)
	DestroyPartitionTable(device string) error
	IsDiskGPT(device string) bool
}
